### PR TITLE
[CELEBORN-1374] Refactor SortBuffer and PartitionSortedBuffer

### PR DESCRIPTION
**Pull request by SteNicholas**
### What changes were proposed in this pull request?

Refactor `SortBuffer` and `PartitionSortedBuffer` with introduction of `DataBuffer` and `SortBasedDataBuffer`.

### Why are the changes needed?

`SortBuffer` and `PartitionSortedBuffer` is refactored in https://github.com/apache/flink/pull/18505. Celeborn Flink should also refactor `SortBuffer` and `PartitionSortedBuffer` to sync the interface changes in Flink. Meanwhile, `SortBuffer` and `PartitionSortedBuffer` should distinguish channel and subpartition for https://github.com/apache/flink/pull/23927.

### Does this PR introduce _any_ user-facing change?

- `SortBuffer` renames to `DataBuffer`.
- `PartitionSortedBuffer` renames to `SortBasedDataBuffer`.
- `SortBuffer.BufferWithChannel` renames to `BufferWithSubpartition`

### How was this patch tested?

UT and IT.

### CODE REVIEW

**Comment by SteNicholas** [comment]
@FMX, @RexXiong, @TanYuxin-tyx.

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/buffer/SortBasedDataBuffer.java [hunk 1]
```diff
@@ -129,14 +130,14 @@ public PartitionSortedBuffer(
...
       }
     }
   }
 
   @Override
-  public boolean append(ByteBuffer source, int targetChannel, DataType dataType)
+  public boolean append(ByteBuffer source, int targetSubpartition, DataType dataType)
       throws IOException {
     checkArgument(source.hasRemaining(), "Cannot append empty data.");
     checkState(!isFinished, "Sort buffer is already finished.");
```
Comment (TanYuxin-tyx) [inline client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/buffer/SortBasedDataBuffer.java:143]: ditto

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/buffer/SortBasedDataBuffer.java [hunk 2]
```diff
@@ -129,14 +130,14 @@ public PartitionSortedBuffer(
...
+        this.subpartitionReadOrder[subpartition] = subpartition;
       }
     }
   }
 
   @Override
-  public boolean append(ByteBuffer source, int targetChannel, DataType dataType)
+  public boolean append(ByteBuffer source, int targetSubpartition, DataType dataType)
       throws IOException {
     checkArgument(source.hasRemaining(), "Cannot append empty data.");
```
Comment (TanYuxin-tyx) [inline client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/buffer/SortBasedDataBuffer.java:142]: Here the message should also be renamed.

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/buffer/SortBasedDataBuffer.java [hunk 3]
```diff
@@ -262,15 +263,15 @@ private void updateWriteSegmentIndexAndOffset(int numBytes) {
   }
 
   @Override
-  public BufferWithChannel copyIntoSegment(
-      MemorySegment target, BufferRecycler recycler, int offset) {
+  public BufferWithSubpartition getNextBuffer(
+      MemorySegment transitBuffer, BufferRecycler recycler, int offset) {
     checkState(hasRemaining(), "No data remaining.");
     checkState(isFinished, "Should finish the sort buffer first before coping any data.");
```
Comment (TanYuxin-tyx) [inline client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/buffer/SortBasedDataBuffer.java:269]: Many other places in the comments show `sort buffer` or `Sort buffer`, we can fix them together.

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleResultPartitionDelegation.java [hunk 4]
```diff
@@ -246,17 +247,17 @@ public void broadcast(ByteBuffer record, Buffer.DataType dataType) throws IOExce
...
+    if (dataBuffer != null) {
+      dataBuffer.release();
     }
   }
 
   public void finish() throws IOException {
     Utils.checkState(
-        unicastSortBuffer == null || unicastSortBuffer.isReleased(),
+        unicastDataBuffer == null || unicastDataBuffer.isReleased(),
         "The unicast sort buffer should be either null or released.");
```
Comment (TanYuxin-tyx) [inline client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleResultPartitionDelegation.java:259]: `data buffer` here

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleResultPartitionDelegation.java [hunk 5]
```diff
@@ -127,70 +128,70 @@ public void emit(
...
-          int subpartitionIndex = bufferWithChannel.getChannelIndex();
-          statisticsConsumer.accept(bufferWithChannel, isBroadcast);
+          Buffer buffer = bufferWithSubpartition.getBuffer();
+          int subpartitionIndex = bufferWithSubpartition.getSubpartitionIndex();
+          statisticsConsumer.accept(bufferWithSubpartition, isBroadcast);
           writeCompressedBufferIfPossible(buffer, subpartitionIndex);
         }
         outputGate.regionFinish();
       } catch (InterruptedException e) {
         throw new IOException("Failed to flush the sort buffer, broadcast=" + isBroadcast, e);
```
Comment (TanYuxin-tyx) [inline client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleResultPartitionDelegation.java:191]: `data buffer`

File: client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleResultPartitionDelegation.java [hunk 6]
```diff
@@ -105,20 +106,20 @@ public void emit(
...
-    if (sortBuffer.append(record, targetSubpartition, dataType)) {
+    DataBuffer dataBuffer = isBroadcast ? getBroadcastDataBuffer() : getUnicastDataBuffer();
+    if (dataBuffer.append(record, targetSubpartition, dataType)) {
       return;
     }
 
     try {
-      if (!sortBuffer.hasRemaining()) {
+      if (!dataBuffer.hasRemaining()) {
         // the record can not be appended to the free sort buffer because it is too large
```
Comment (TanYuxin-tyx) [inline client-flink/common/src/main/java/org/apache/celeborn/plugin/flink/RemoteShuffleResultPartitionDelegation.java:116]: `data buffer`

**Review by TanYuxin-tyx — COMMENTED** [review]
@SteNicholas The change looks good to me overall. I only left some minor comments about the docs or typos. PTAL, thanks.

**Review by TanYuxin-tyx — APPROVED** [review]
Thanks for the update. LGTM.

**Review by RexXiong — APPROVED** [review]
LGTM, thanks!

**Pull request closed without merging**